plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/GitEventsHandler.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.ide.ext.git.client;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.util.HashSet;
+import java.util.Set;
+import org.eclipse.che.api.core.jsonrpc.commons.RequestHandlerConfigurator;
+import org.eclipse.che.api.git.shared.FileChangedEventDto;
+import org.eclipse.che.api.git.shared.StatusChangedEventDto;
+import org.eclipse.che.api.project.shared.dto.event.GitCheckoutEventDto;
+
+/**
+ * Receives git events from server side.
+ *
+ * @author Igor Vinokur
+ * @author Mykola Morhun
+ */
+@Singleton
+public class GitEventsHandler implements GitEventSubscribable {
+
+  private static final String EVENT_GIT_FILE_CHANGED = "event/git-change";
+  private static final String EVENT_GIT_STATUS_CHANGED = "event/git/statusChanged";
+  private static final String EVENT_GIT_CHECKOUT = "event/git-checkout";
+
+  private final Set<GitEventsSubscriber> subscribers = new HashSet<>();
+
+  @Inject
+  public GitEventsHandler(RequestHandlerConfigurator configurator) {
+    configureHandlers(configurator);
+  }
+
+  private void configureHandlers(RequestHandlerConfigurator configurator) {
+    configurator
+        .newConfiguration()
+        .methodName(EVENT_GIT_FILE_CHANGED)
+        .paramsAsDto(FileChangedEventDto.class)
+        .noResult()
+        .withBiConsumer(this::onFileChangedHandler);
+
+    configurator
+        .newConfiguration()
+        .methodName(EVENT_GIT_STATUS_CHANGED)
+        .paramsAsDto(StatusChangedEventDto.class)
+        .noResult()
+        .withBiConsumer(this::onStatusChangedHandler);
+
+    configurator
+        .newConfiguration()
+        .methodName(EVENT_GIT_CHECKOUT)
+        .paramsAsDto(GitCheckoutEventDto.class)
+        .noResult()
+        .withBiConsumer(this::onCheckoutHandler);
+  }
+
+  private void onFileChangedHandler(String endpointId, FileChangedEventDto fileChangedEventDto) {
+    for (GitEventsSubscriber subscriber : subscribers) {
+      subscriber.onFileChanged(endpointId, fileChangedEventDto);
+    }
+  }
+
+  private void onStatusChangedHandler(
```

Comment:
Have you considered using GWT's EventBus? Is there any reason not to use it here?
As far as I see all this code duplicates EventBus's functionality.

Reply:
Agree with Artem

Reply:
Potential problem with using event bus here is heavy load of the `GitEventsHandler`. It fires an event on each editing of a file, which could cause some undesirable delays in IDE.

Reply:
yeah, indeed, it may really heavy load for EventBus. In this case, I agree that using separate events dispatching mechanism will be better.